File: src/core/accumulator.ts
import type { TraceArg } from './propagation.ts';
import type { WideEvent } from './schema';
import type { BeginMeta } from './types.ts';

import { normalize } from './normalize.ts';
import { deepMerge, readString } from './wide-event.ts';

export type BeginChild = (meta: BeginMeta) => Event;
export type Emit = (event: WideEvent) => void;

const perf: undefined | { now?: () => number } = globalThis.performance;

export class Event {
	get spanId(): string | undefined {
		return readString(this.#data.span_id);
	}
	get traceId(): string | undefined {
		return readString(this.#data.trace_id);
	}
	readonly #beginChild: BeginChild;
	readonly #data: WideEvent;
	readonly #emit: Emit;
	#ended = false;
	readonly #startedAt: number;

	constructor(emit: Emit, event: WideEvent, beginChild: BeginChild) {
		this.#emit = emit;
		this.#data = event;
		this.#beginChild = beginChild;
		this.#startedAt = monotonicNow();
	}

	add = (fields: Partial<WideEvent>): this => {
		if (this.#ended) {
			return this;
		}

		deepMerge(this.#data, fields);
		return this;
	};

	child = (event: string): Event => {
		return this.#beginChild({ event, parent_span_id: this.spanId, trace_id: this.traceId });
	};

	end = (): void => {
		if (this.#ended) return;
		this.#ended = true;
		this.#data.duration_ms = Math.max(0, Math.round(monotonicNow() - this.#startedAt));
		normalize(this.#data);
		this.#emit(this.#data);
	};

	error = (err: unknown, fields?: Partial<WideEvent>): this => {
		if (this.#ended) {
			return this;
		}

		if (fields !== undefined) {
			deepMerge(this.#data, fields);
		}
		this.#data.outcome = 'error';
		this.#data.error = toErrorFields(err);
		return this;
	};

	// Total because begin() assigns both ids unconditionally when constructing the
	// envelope; a future Event path that defers id assignment would reintroduce
	// undefined behind this signature and break required-trace callees.
	traceArg = (): TraceArg => {
		return { parent_span_id: this.#data.span_id, trace_id: this.#data.trace_id };
	};
}

function monotonicNow(): number {
	return perf?.now === undefined ? Date.now() : perf.now();
}

function readErrorCode(err: object): string | undefined {
	if ('code' in err && typeof err.code === 'string') {
		return err.code;
	}

	if ('data' in err && err.data !== null && typeof err.data === 'object' && 'code' in err.data) {
		return typeof err.data.code === 'string' ? err.data.code : undefined;
	}

	return undefined;
}

function toErrorFields(err: unknown): { code?: string; message?: string; type: string } {
	const type = err instanceof Error ? err.name : 'UnknownError';
	const message = err instanceof Error ? err.message : String(err);
	const code = err !== null && typeof err === 'object' ? readErrorCode(err) : undefined;
	if (code !== undefined) {
		return { code, message, type };
	}

	return { message, type };
}

File: python/observe_py/accumulator.py
"""Span accumulator — twin of the ``Event`` class in ``src/core/accumulator.ts``.

Named ``Span`` per ``docs/python-port.md``; ``end()`` runs normalize and hands the
bag to the client's emit pipeline exactly once.
"""

from __future__ import annotations

import time
from collections.abc import Callable
from typing import Any

from .normalize import normalize
from .wide_event import deep_merge, read_string

BeginChild = Callable[..., "Span"]
Emit = Callable[[dict[str, Any]], None]


class Span:
    def __init__(self, emit: Emit, data: dict[str, Any], begin_child: BeginChild) -> None:
        self._emit = emit
        self._data = data
        self._begin_child = begin_child
        self._ended = False
        self._started_at = time.monotonic()

    @property
    def span_id(self) -> str | None:
        return read_string(self._data.get("span_id"))

    @property
    def trace_id(self) -> str | None:
        return read_string(self._data.get("trace_id"))

    def add(self, fields: dict[str, Any]) -> Span:
        if self._ended:
            return self

        deep_merge(self._data, fields)
        return self

    def child(self, event: str) -> Span:
        return self._begin_child(event, parent_span_id=self.span_id, trace_id=self.trace_id)

    def end(self) -> None:
        if self._ended:
            return
        self._ended = True
        self._data["duration_ms"] = round((time.monotonic() - self._started_at) * 1000, 3)
        normalize(self._data)
        self._emit(self._data)

    def error(self, err: object, fields: dict[str, Any] | None = None) -> Span:
        if self._ended:
            return self

        if fields is not None:
            deep_merge(self._data, fields)
        self._data["outcome"] = "error"
        self._data["error"] = _to_error_fields(err)
        return self


def _to_error_fields(err: object) -> dict[str, str]:
    if isinstance(err, BaseException):
        kind = type(err).__name__
        message = str(err)
    else:
        kind = "UnknownError"
        message = str(err)

    fields = {"message": message, "type": kind}
    code = getattr(err, "code", None)
    if not isinstance(code, str):
        data = getattr(err, "data", None)
        if isinstance(data, dict):
            code = data.get("code")
    if isinstance(code, str):
        fields["code"] = code

    return fields
